### ISSUE TITLE
Deadlocks detected on DBLockingProvider.php after update to 10.5

### ISSUE THREAD

**Issue by mojontwins** [owncloud/core#38221]
### Steps to reproduce
1. Seems to happen randomly

### Expected behaviour
No Deadlocks detected, like with previous versions.

### Actual behaviour
We've been maintaining an Owncloud installation and upgrading to each version since v9 without problems. We were running 10.2 for a while flawlessly but as soon as we upgraded to 10.5 we are getting deadlocks errors reported in the apache `error_log` file:

```
[Mon Dec 14 08:33:58.148407 2020] [php7:error] [pid 87941] [client 10.226.250.13:58662] PHP Fatal error:  Uncaught PDOException: SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction in /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:127
Stack trace:
#0 /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php(127): PDOStatement->execute(NULL)
#1 /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1062): Doctrine\\DBAL\\Driver\\PDOStatement->execute()
#2 /var/www/html/owncloud/lib/private/DB/Connection.php(207): Doctrine\\DBAL\\Connection->executeUpdate('UPDATE `oc_file...', Array, Array)
#3 /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php(209): OC\\DB\\Connection->executeUpdate('UPDATE `oc_file...', Array, Array)
#4 /var/www/html/owncloud/lib/private/DB/QueryBuilder/QueryBuilder.php(141): Doctrine\\DBAL\\Query\\QueryBuilder->execute()
#5 /var/www/html/owncloud/lib/private/Lock/DBLockingProvider.php(281): OC\\DB\\QueryB in /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php on line 34
```

### Server configuration
**Operating system**: RHEL 7

**Web server:** Apache 2.4

**Database:** MySQL 5.7

**PHP version:** 7.2

**ownCloud version:** 10.5

**Updated from an older ownCloud or fresh install:** Updated from 10.2

**Where did you install ownCloud from:** .tar.gz package, manual install

**Signing status (ownCloud 9.0 and above):**

```
https://gist.github.com/mojontwins/6dac4d5b61b73ea7eba6d39d1d54a9a1
```

**The content of config/config.php:**

```
{
    "system": {
        "updatechecker": false,
        "instanceid": "ocd0oa4vyocy",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "10.240.243.167"
        ],
        "datadirectory": "\/cloud",
        "overwrite.cli.url": "http:\/\/10.240.243.167",
        "dbtype": "mysql",
        "version": "10.5.0.10",
        "dbname": "owncloud",
        "dbhost": "10.240.243.178",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "ldapIgnoreNamingRules": false,
        "theme": "",
        "log_type": "owncloud",
        "logfile": "\/var\/log\/owncloud.log",
        "logdateformat": "F d, Y H:i:s",
        "default_language": "es",
        "loglevel": 1,
        "maintenance": false,
        "htaccess.RewriteBase": "\/",
        "token_auth_enforced": false,
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/owncloud\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/owncloud\/apps-external",
                "url": "\/apps-external",
                "writable": true
            }
        ],
        "enable_avatars": true,
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": 25,
        "session_lifetime": 1800,
        "session_keepalive": false
    }
}

```

**List of activated apps:**

Enabled:
  - brute_force_protection: 1.0.1
  - comments: 0.3.0
  - configreport: 0.2.0
  - dav: 0.6.0
  - federatedfilesharing: 0.5.0
  - files: 1.5.2
  - files_external: 0.7.1
  - files_mediaviewer: 1.0.3
  - files_pdfviewer: 0.11.2
  - files_sharing: 0.13.0
  - files_texteditor: 2.3.0
  - files_trashbin: 0.9.1
  - files_versions: 1.3.0
  - files_videoplayer: 0.10.1
  - firstrunwizard: 1.2.0
  - market: 0.5.0
  - notifications: 0.5.2
  - provisioning_api: 0.5.0
  - systemtags: 0.3.0
  - theme-jda: 1.0.0
  - updatenotification: 0.2.1
  - user_ldap: 0.15.2
  - user_saml: 0.4.0
Disabled:
  - activity
  - admin_audit
  - announcementcenter
  - customgroups
  - encryption
  - enterprise_key
  - external
  - federation
  - files_antivirus
  - files_classifier
  - files_external_dropbox
  - files_external_ftp
  - files_ldap_home
  - firewall
  - guests
  - oauth2
  - password_policy
  - ransomware_protection
  - sharepoint
  - systemtags_management
  - templateeditor
  - theme-enterprise
  - twofactor_totp
  - user_external
  - user_shibboleth
  - windows_network_drive
  - wopi
  - workflow

```

**Are you using external storage, if yes which one:** nfs.

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP

#### LDAP configuration (delete this part if not used)

```
+-------------------------------+----------------------------------------------------------------------------------+
| Configuration                 | s01                                                                              |
+-------------------------------+----------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      | 0                                                                                |
| hasPagedResultSupport         |                                                                                  |
| homeFolderNamingRule          | attr:uid                                                                         |
| lastJpegPhotoLookup           | 0                                                                                |
| ldapAgentName                 |                                                                                  |
| ldapAgentPassword             | ***                                                                              |
| ldapAttributesForGroupSearch  |                                                                                  |
| ldapAttributesForUserSearch   | uid;cn                                                                           |
| ldapBackupHost                |                                                                                  |
| ldapBackupPort                |                                                                                  |
| ldapBase                      | o=empleados,o=juntadeandalucia,c=es                                              |
| ldapBaseGroups                | o=empleados,o=juntadeandalucia,c=es                                              |
| ldapBaseUsers                 | o=empleados,o=juntadeandalucia,c=es                                              |
| ldapCacheTTL                  | 600                                                                              |
| ldapConfigurationActive       | 1                                                                                |
| ldapDynamicGroupMemberURL     |                                                                                  |
| ldapEmailAttribute            | mail                                                                             |
| ldapExperiencedAdmin          | 1                                                                                |
| ldapExpertUUIDGroupAttr       |                                                                                  |
| ldapExpertUUIDUserAttr        | uid                                                                              |
| ldapExpertUsernameAttr        | uid                                                                              |
| ldapGroupDisplayName          | cn                                                                               |
| ldapGroupFilter               |                                                                                  |
| ldapGroupFilterGroups         |                                                                                  |
| ldapGroupFilterMode           | 0                                                                                |
| ldapGroupFilterObjectclass    |                                                                                  |
| ldapGroupMemberAssocAttr      | uniqueMember                                                                     |
| ldapHost                      | ldap23.juntadeandalucia.es                                                       |
| ldapIgnoreNamingRules         |                                                                                  |
| ldapLoginFilter               | (&(objectclass=empleado)(|(uid=%uid)(mail=%uid))(JAcloud=SI)(!(passCaducado=1))) |
| ldapLoginFilterAttributes     |                                                                                  |
| ldapLoginFilterEmail          | 0                                                                                |
| ldapLoginFilterMode           | 0                                                                                |
| ldapLoginFilterUsername       | 1                                                                                |
| ldapNestedGroups              | 0                                                                                |
| ldapNetworkTimeout            | 2                                                                                |
| ldapOverrideMainServer        |                                                                                  |
| ldapPagingSize                | 500                                                                              |
| ldapPort                      | 389                                                                              |
| ldapQuotaAttribute            | JAcloudQuota                                                                     |
| ldapQuotaDefault              |                                                                                  |
| ldapTLS                       | 0                                                                                |
| ldapUserDisplayName           | cn                                                                               |
| ldapUserDisplayName2          |                                                                                  |
| ldapUserFilter                | (&(objectclass=empleado)(JAcloud=SI)(!(passCaducado=1)))                         |
| ldapUserFilterGroups          |                                                                                  |
| ldapUserFilterMode            | 0                                                                                |
| ldapUserFilterObjectclass     | inetOrgPerson                                                                    |
| ldapUserName                  | samaccountname                                                                   |
| ldapUuidGroupAttribute        | auto                                                                             |
| ldapUuidUserAttribute         | auto                                                                             |
| turnOffCertCheck              | 0                                                                                |
| useMemberOfToDetectMembership | 1                                                                                |
+-------------------------------+----------------------------------------------------------------------------------+

```

### Client configuration
**Browser:** N/A

**Operating system:** N/A

### Logs
#### Web server error log
```
[Mon Dec 14 08:33:58.148407 2020] [php7:error] [pid 87941] [client 10.226.250.13:58662] PHP Fatal error:  Uncaught PDOException: SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction in /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:127\nStack trace:\n#0 /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php(127): PDOStatement->execute(NULL)\n#1 /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1062): Doctrine\\DBAL\\Driver\\PDOStatement->execute()\n#2 /var/www/html/owncloud/lib/private/DB/Connection.php(207): Doctrine\\DBAL\\Connection->executeUpdate('UPDATE `oc_file...', Array, Array)\n#3 /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php(209): OC\\DB\\Connection->executeUpdate('UPDATE `oc_file...', Array, Array)\n#4 /var/www/html/owncloud/lib/private/DB/QueryBuilder/QueryBuilder.php(141): Doctrine\\DBAL\\Query\\QueryBuilder->execute()\n#5 /var/www/html/owncloud/lib/private/Lock/DBLockingProvider.php(281): OC\\DB\\QueryB in /var/www/html/owncloud/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php on line 34
```

#### ownCloud log (data/owncloud.log)
```
Nothing about this issue in owncloud.log
```

#### Browser log
```
N/A
```

**Comment by mojontwins** [comment]
We managed to have the database give us more info:

```
------------------------
LATEST DETECTED DEADLOCK
------------------------
2021-02-01 10:47:40 0x7eff51b52700
*** (1) TRANSACTION:
TRANSACTION 405802495, ACTIVE 0 sec starting index read
mysql tables in use 1, locked 1
LOCK WAIT 4 lock struct(s), heap size 1136, 3 row lock(s), undo log entries 1
MySQL thread id 26176412, OS thread handle 139635058804480, query id 1622549861 10.226.133.204 owncloud updating
UPDATE `oc_file_locks` SET `lock` = `lock` -1 WHERE (`key` IN ('files/eaf4c4e5ce126a82c2eb9e63a2f7a7fc', 'files/16d8d3e854598b749ddc31f8ca0e115f', 'files/1f29ba865302344914cf335329531e6e', 'files/45e0cfc2f66b74c5bba4fbcd0baf31de', 'files/2eafc67c75f50655af4efcfd71e557dc', 'files/3c7e103364f9f356feb6d10d48aeb16f', 'files/b983b68ecae5327dc7e6f2aa8b83bb1e', 'files/079429a4cd3175aec550a4076f042dd2', 'files/bfde4f1d96c199f2f05756473c4f1ec6', 'files/9b2adee2c4a7427f9b8e289601330cdd', 'files/e4be36dbefa26118293644fd8fcbad1c', 'files/0949bc92f195cb2e42019d02afb53e45', 'files/8dde4fb2272ef89050b13a35c0bf3b8d', 'files/8aea4ae89fc713fce942346c87dab5e1', 'files/8175ce56ed960daee23183b52a29d26c', 'files/7bce73a6b60bcefb91ad8c0410e85952', 'files/fcbb0b5b72340b27a110ceeb03a90039', 'files/0ae5c2b03d2b23ced216794d4efc0ad9', 'files/5f02f74f4deba5584ef887c6f9196752', 'files/e99e0f9c0c047dc7c09060a543a4b110', 'files/
*** (1) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 269 page no 8415 n bits 344 index lock_key_index of table `owncloud`.`oc_file_locks` trx id 405802495 lock_mode X locks rec but not gap waiting
Record lock, heap no 247 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 30; hex 66696c65732f303739343239613463643331373561656335353061343037; asc files/079429a4cd3175aec550a407; (total 38 bytes);
 1: len 4; hex 003f6b93; asc  ?k ;;

*** (2) TRANSACTION:
TRANSACTION 405802494, ACTIVE 0 sec starting index read
mysql tables in use 1, locked 1
37 lock struct(s), heap size 8400, 69 row lock(s), undo log entries 34
MySQL thread id 26176414, OS thread handle 139635052586752, query id 1622549929 10.226.133.205 owncloud updating
UPDATE `oc_file_locks` SET `lock` = `lock` + 1, `ttl` = '1612176460' WHERE `key` = 'files/041178a2a12a3738960bdc27e5f69064' AND `lock` >= 0
*** (2) HOLDS THE LOCK(S):
RECORD LOCKS space id 269 page no 8415 n bits 344 index lock_key_index of table `owncloud`.`oc_file_locks` trx id 405802494 lock_mode X locks rec but not gap
Record lock, heap no 247 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 30; hex 66696c65732f303739343239613463643331373561656335353061343037; asc files/079429a4cd3175aec550a407; (total 38 bytes);
 1: len 4; hex 003f6b93; asc  ?k ;;

*** (2) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 269 page no 5655 n bits 408 index lock_key_index of table `owncloud`.`oc_file_locks` trx id 405802494 lock_mode X locks rec but not gap waiting
Record lock, heap no 252 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 30; hex 66696c65732f303431313738613261313261333733383936306264633237; asc files/041178a2a12a3738960bdc27; (total 38 bytes);
 1: len 4; hex 003f6bcd; asc  ?k ;;

*** WE ROLL BACK TRANSACTION (1)
```

As you see, the massive "set lock = -1" query from `lib/private/Lock/DBLockingProvider.php` gets a dead lock with a `SET `lock` = `lock` + 1` from elsewhere.

This makes the first query to fail and leaves the files locked and the user account unusable for a while.

**Comment by mojontwins** [comment]
Hello, more info. I don't know if this is related, but when sometimes users get a yellow notice "Failed to request notifications. Please reload the page manually", and in the Javascript console I can see HTTP 401 errors, which didn't happen with the previous version 10.2 before upgrading to 10.5

![image](https://user-images.githubusercontent.com/6029672/107474243-4d2a6580-6b72-11eb-869b-f4350180553b.png)

**Comment by DeepDiver1975** [comment]
**Classification: confirmed-bug** (analysis only — no automated fix PR)

🤖 Automated issue analysis by Claude Code review agent.

Thanks for the detailed InnoDB `LATEST DETECTED DEADLOCK` dump — it pinpoints the root cause precisely, and the deadlock-prone code is **still present on current `master`** (`lib/private/Lock/DBLockingProvider.php`).

### Root cause

The two transactions in your dump are both `DBLockingProvider` queries that take `lock_mode X` record locks on `lock_key_index` of `oc_file_locks`, in **opposite key orders**, so they can wait on each other:

- **TX (2)** — `acquireLock()` taking a shared lock, one key at a time:
  ```sql
  UPDATE `oc_file_locks` SET `lock` = `lock` + 1, `ttl` = ? WHERE `key` = ? AND `lock` >= 0
  ```
  (master `acquireLock()`, the `LOCK_SHARED` branch)
- **TX (1)** — `releaseAll()` decrementing many keys in **one** statement:
  ```sql
  UPDATE `oc_file_locks` SET `lock` = `lock` - 1 WHERE (`key` IN (...up to 100 keys...)) AND `lock` > 0
  ```
  (master `releaseAll()`, which chunks shared-lock release into batches of 100 via `array_chunk(..., 100)`)

Because `releaseAll()` locks a *batch* of index records (in the order MySQL walks the `IN` list / index) while concurrent `acquireLock()` calls each lock a *single* record, two requests can acquire the same two records in different orders → classic InnoDB deadlock 1213 / `SQLSTATE[40001]`. The losing transaction is rolled back, surfacing as the `Uncaught PDOException` in your Apache `error_log` (your stack frame `DBLockingProvider.php(281)`), and leaving the user's files locked until the TTL expires.

### Why it appeared after 10.2 → 10.5

This isn't new code in 10.5, but throughput/concurrency characteristics changed (more parallel desktop-client and sync traffic per request), making the pre-existing race observable. The `notifications`/HTTP 401 you mention is almost certainly unrelated to the deadlock (separate auth/notification path) and not part of this report.

### Current state on master

- `acquireLock()` / `releaseLock()` / `changeLock()` / `releaseAll()` all run their `UPDATE`s via `IDBConnection::executeUpdate()` **with no deadlock catch-and-retry**.
- The only deadlock-aware retry in the codebase is `OC\DB\Adapter::upsert()` (`catch (DeadlockException $e) { ... retry }`), which is a **different** code path and does **not** wrap the locking-provider statements. So a deadlock here is fatal rather than transparently retried.

### Suggested directions (for maintainers)

1. **Wrap the locking-provider UPDATEs in a bounded deadlock-retry** (mirroring `Adapter::upsert()`'s `DeadlockException` handling) so a 1213 is retried instead of bubbling up as a fatal `PDOException`.
2. **Reduce lock-ordering conflicts in `releaseAll()`** — e.g. sort keys deterministically before the batched `IN` UPDATE, and/or shrink the chunk size, so all sessions acquire record locks in a consistent order.
3. **Operational workaround for affected admins:** switch to the **Redis (memcache) locking provider** (`'filelocking.enabled' => true` with `'memcache.locking' => '\\OC\\Memcache\\Redis'`), which avoids `oc_file_locks` row contention entirely. This is the recommended configuration for busy multi-server installs and sidesteps the DB deadlock.

This is related to the same `DBLockingProvider` contention area as #38207 and #38039.

(Confirmed against `master`; this is an automated analysis comment, no code change is being submitted.)

**Comment by dj4oC** [comment]
## Implementation prompt

**Problem:** `lib/private/Lock/DBLockingProvider.php`'s `acquireLock()` (single-key `UPDATE ... lock = lock + 1`) and `releaseAll()` (batched, up-to-100-key `UPDATE ... lock = lock - 1 WHERE key IN (...)`) can lock the same `oc_file_locks` index records in different orders under concurrency, producing InnoDB deadlocks (error 1213). Unlike `OC\DB\Adapter::upsert()`, which already retries on `DeadlockException`, none of the locking-provider UPDATEs (`acquireLock`, `releaseLock`, `changeLock`, `releaseAll`) catch or retry deadlocks — a deadlock is fatal, leaving the user's files effectively locked until TTL expiry.

**Acceptance criteria:**
- A `DeadlockException` (or the underlying driver deadlock error) raised by any `DBLockingProvider` lock/release UPDATE is retried a bounded number of times instead of propagating as an uncaught `PDOException`.
- `releaseAll()`'s batched key list is ordered deterministically (e.g. sorted) before the UPDATE, to reduce lock-ordering conflicts with concurrent single-key `acquireLock()` calls.
- No change to correct lock/release semantics — only deadlock resilience and lock-ordering are affected.

**Files:**
- `lib/private/Lock/DBLockingProvider.php` — `acquireLock()`, `releaseLock()`, `changeLock()`, `releaseAll()`: add bounded deadlock-retry (mirroring the pattern in `lib/private/DB/Adapter.php::upsert()`), and sort the key list in `releaseAll()`'s batched `IN` clause before executing.

**Constraints:**
- Test coverage ≥ 85% on touched code (e2e preferred — run the tool, do not trust claimed %)
- Commit signing: GPG signature + DCO Signed-off-by
- Rebase onto base branch before pushing
- No root-directory edits

🤖 Automated by Claude Code

---
_Generated by [Claude Code](https://claude.ai/code)_
